### ISSUE TITLE
Issues with missing personal files and no access to shared files after upgrading from 6.0.6 to 7.x

### ISSUE THREAD

**Issue by fleish** [owncloud/core#12801]
### Steps to reproduce
1. upgrade owncloud 6.0.6 using SQLite database to 7.0.3 or 7.0.4 using ubuntu package from http://download.opensuse.org/repositories/isv:/ownCloud:/community:/7.0/xUbuntu_12.04/
2. run occ upgrade
3. (optional - issue occurs in either database type) convert from SQLite to MySQL using occ db:convert-type --all-apps mysql owncloud localhost 
4. Login to your owncloud via the web

### Expected behaviour
You should be able to see your files & the files that have been shared with you

### Actual behaviour
You will not be able to view your files nor any of the files that have been shared with you (for the latter you can see the folders that have been shared with you but you cannot get into them)

NOTE: talked to AnybodyElse on IRC about this and as suggested ran "occ files:scan --all" which fixed the first issue of not being able to see my own files. 

### Server configuration
**Operating system**: Ubuntu 12.04.5 LTS

**Web server:** Apache/2.2.22 (Ubuntu)

**Database:** SQLite3 or mysql  Ver 14.14 Distrib 5.5.40, for debian-linux-gnu (x86_64) using readline 6.2

**PHP version:** PHP 5.3.10-1ubuntu3.15 with Suhosin-Patch (cli) (built: Oct 29 2014 12:19:04) 

**ownCloud version:** 7.0.4.2

**Updated from an older ownCloud or fresh install:** updated from 6.0.6

**List of activated apps:** 
Activity
Deleted Files
Documents
Encryption
First Run Wizard
Full Text Search
LDAP user and group backend
PDF Viewer
Pictures
Share Files
Text Editor
Video Viewer

**The content of config/config.php:**
<?php
$CONFIG = array (
  'instanceid' => '**REDACTED**',
  'passwordsalt' => '**REDACTED**',
  'datadirectory' => '/var/www/owncloud/data',
  'dbtype' => 'mysql',
  'version' => '7.0.4.2',
  'installed' => true,
  'ldapIgnoreNamingRules' => false,
  'forcessl' => true,
  'enable_previews' => false,
  'theme' => '',
  'maintenance' => false,
  'trusted_domains' => 
  array (
    0 => '**REDACTED**',
  ),
  'secret' => '**REDACTED**',
  'share_folder' => '/Shared',
  'dbname' => 'owncloud',
  'dbhost' => 'localhost',
  'dbuser' => 'owncloud',
  'dbpassword' => '**REDACTED**',
);

**Are you using external storage, if yes which one:** NO

**Are you using encryption:** YES

### Client configuration
**Browser:** Google Chrome Version 39.0.2171.71 (64-bit)

**Operating system:** Mac OSX 10.8.5 (also tested in Mac OSX 10.9.5)

#### Web server error log
[Thu Dec 11 21:12:27 2014] [error] [client 50.248.219.14] PHP Fatal error:  Cannot redeclare class OC_Search_Provider in /var/www/owncloud/lib/private/search/provider.php on line 5, referer: https://owncloud2.newconews.org/index.php/apps/files/?dir=%2F

#### ownCloud log (data/owncloud.log)
messages seen after the upgrade:
1) This one appears repeatedly. I think it corresponds to me trying to view things that have been shared with me by user with UUID yyy-yyy-yyy-yyy (REDACTED):
{"app":"files_sharing","message":"Unable to get mount for shared storage '3' user 'yyy-yyy-yyy-yyy'","level":3,"time":"2014-12-11T21:07:31+00:00"} 

2) This one appears just once, seems to happen when trying to view shared files. First error No. 1 is repeated twice, then this message appears, then error No. 1 is repeated many more times:
{"app":"PHP","message":"Invalid argument supplied for foreach() at \/var\/www\/owncloud\/lib\/private\/files\/cache\/scanner.php#216","level":3,"time":"2014-12-11T21:07:31+00:00"}

3) This one appears a couple of times:
{"app":"PHP","message":"Cannot redeclare class OC_Search_Provider at \/var\/www\/owncloud\/lib\/private\/search\/provider.php#5","level":3,"time":"2014-12-11T21:12:27+00:00"}

4) These appeared once in conjunction with me deleting a file on my computer (REDACTED my UUID xxx-xxx-xxx-xxx) and my sync client pushing it to the server. The file was actually deleted, but the messages were logged:
{"app":"Encryption library","message":"No share key found for user \"xxx-xxx-xxx-xxx\" for file \"Test222.txt\"","level":2,"time":"2014-12-11T21:11:47+00:00"}
{"app":"core","message":"No preview providers exist","level":3,"time":"2014-12-11T21:11:47+00:00"}
{"app":"hook","message":"error while running hook (OC\\Preview::post_delete_files): No preview providers","level":3,"time":"2014-12-11T21:11:47+00:00"}

**Comment by MorrisJobke** [comment]
@fleish Is the instance working after the upgrade and before the database migration? (after step 2) Or is it still broken at this point. I ask this to find the cause of your problem (either upgrade code or migration code)

**Comment by fleish** [comment]
@MorrisJobke sorry if that was unclear, I tried to caveat step #3 to note it happens regardless of the DB migration. I've tried just upgrading the code and leaving the instance on SQLite and the same issue(s) occur

**Comment by fleish** [comment]
@MorrisJobke or anyone else able to provide some suggestions here? My users are getting restless with daily issues with sqlite locking

**Comment by MorrisJobke** [comment]
@PVince81 @schiesbn YOu have more experience with the sharing part and maybe can better analyze this one. Thanks

**Comment by PVince81** [comment]
@fleish please check your `oc_storages` table.
Are there entries with the format "local::$datadir/$user" and also "home::$user" ? (there should be only one per user, not both). Did you move/change/rename the data dir when upgrading ?

**Comment by fleish** [comment]
@PVince81 checking in oc_storages pre-upgrade I see only 1 local:: entry as follows:
local::/var/www/owncloud/data/|4

the rest are like:
home::admin|1
home::**REDACTED-GUID**|2

I don't change anything manually during the upgrade, only whatever the apt/dpkg upgrade does and then occ upgrade & occ files:scan --all

**Comment by PVince81** [comment]
Okay good, so your setup isn't a case of the known "home storage duplication" which is known to make shares disappear.

**Comment by PVince81** [comment]
Let's check for orphaned shares:
`select * from oc_share where item_source not in (select fileid from oc_filecache)`

and orphaned storages:
`select * from oc_filecache where storage not in (select numeric_id from oc_storages)`

Also that error 3) about the search provider class is suspicious. Do you somehow happen to have the search app enabled twice from different sources ?

**Comment by fleish** [comment]
Ophaned shares:
19|**REDACTED_GUID1**|**REDACTED_GUID2**||file|1721|/1721|1721|/REDACTED_FILENAME 3.zip.gpg|||0|27|1398283250|0|0
20|**REDACTED_GUID3**|**REDACTED_GUID2**||file|1721|/1721|1721|/REDACTED_FILENAME 3.zip.gpg|||0|27|1398283261|0|0
85|**REDACTED_GUID4**|**REDACTED_GUID5**||file|34047|/34047|34047|/REDACTED_FILENAME.pdf|||0|27|1409772053|0|0

Nothing for orphaned storages.

When I check the web UI for apps, I only see Full Text Search once in the list. I'm not sure how else to check for this case.

**Comment by PVince81** [comment]
I see the error above says error with shared storage 3.

Is there a storage with this numeric_id in the oc_storage?
`select * from oc_storages where numeric_id=3`

**Comment by fleish** [comment]
Yes, just one
home::**REDACTED_GUID**|3

**Comment by fleish** [comment]
anything else we can try to debug this one?

**Comment by PVince81** [comment]
Ok let's try and find the shared files that do not appear:
`select * from oc_share where file_target like '%FILENAME%'` (replace FILENAME with a part of the file or folder name that appears in "Shared with you" but not "All files".
Please post the result and check whether "owner" and "shared_with" fields are correct.
Then, write down the matching "item_source", and look up the file id for it:

`select * from oc_filecache where fileid=THEITEMSOURCE`. Replace THEITEMSOURCE with the item_source id you found in the first query.

**Comment by PVince81** [comment]
You said your users are having "database is locked" issues with sqlite ? How many users do you have ?
If more than 1-2, it is best to switch to mysql (you can convert the database with this tool: http://doc.owncloud.org/server/7.0/admin_manual/maintenance/convert_db.html) to avoid database locks.

**Comment by fleish** [comment]
Just to be clear, I've been running these on the instance with issues that I have rolled back to version 6.

Here are the first 2 items from oc_share. The first one is a folder shared with a group, while the second is a folder shared to an individual user. The GUID & group values look to be correct.

7|IT|**REDACTED_GUID1**||folder|356|/356|356|/IT|||1|31|1393605790|0|0

63|**REDACTED_GUID2**|**REDACTED_GUID3**||folder|7184|/7184|7184|/Some Pathname|||0|31|1403635337|0|1

The corresponding oc_filecache details:
356|3|files/IT|8c9b25373ad9c17c37942f323c355ab0|69|IT|2|1|1421463681|1418193552|0|54b9d0b4b9ad0|564708132|420198229

7184|15|files/documents/Some Pathname|1f8fe0f51b8b8b433d2a65072c8246a9|1573|Some Pathname|2|1|1418404137|1405710069|0|548b212c84f94|3607992|2697910


As for the database locked issues with sqlite. We desperately want to migrate it to mysql, however the DB conversion utility was not introduced until version 7 so we are kind of stuck in a catch 22 until we can get our instance upgraded from 6->7.

**Comment by PVince81** [comment]
You said there is a user Y who doesn't see the files/folders F shared with them in "All files".
Does user Y see the shares in the "Shared with you" section of the sidebar ?

I was wondering whether the scenario of the missing files match the following:

- User U1 shares "folder" with group G1.
- User U2 (member of G1) shares "folder/subfolder" with user U3
- User U3 is unable to see "subfolder"

If your situation matches the scenario, this could be the issue from https://github.com/owncloud/core/issues/12475

**Comment by fleish** [comment]
That scenario doesn't match. Clicking on "shared with you" in the sidebar shows the folders that are shared with you ... but trying to access any of them the window just spins for a bit and then redirects to "index.php/apps/files/?dir=%2F"

And owncloud.log shows this where REDACTED_UID1 is the user who shared the files with me
{"app":"files_sharing","message":"Unable to get mount for shared storage '3' user '**REDACTED_UID1**'","level":3,"time":"2015-01-20T23:12:01+00:00"}

**Comment by PVince81** [comment]
Hmmm, but your symptoms match the other issue. You could still try with the patch https://github.com/owncloud/core/pull/13470

**Comment by PVince81** [comment]
After applying the patch, the user who reshared with the other one needs to unshare+reshare to make it work.

**Comment by fleish** [comment]
To confirm, this is a patch for v7? If so, do I apply it after completing the upgrade from 6 to 7? Or after the new files are present but before the occ upgrade is run?

**Comment by PVince81** [comment]
It doesn't matter when you apply it.
Once applied, it will affect newly created reshares.
So if you had broken reshares you need to unshare + reshare them.

**Comment by fleish** [comment]
I don't think we have any re-shares so hopefully unsharing/re-sharing won't be required. If that ends up being the case we might as well nuke the whole instance and just start over. I don't see a tests sub-dir for that second file in the patch. Is it just ./lib/private/share/share.php that needs updating?

**Comment by PVince81** [comment]
Yes, only update "lib/private/share/share.php".

If you didn't have any reshares the fix might still work because it affects the logic that checks "give me all shares that target user XYZ through his groups" (the broken part was "through his groups if the user has more than 1 group assigned"). I'd say try it out with a few broken shares. If they're still broken after the patch, try recreating them.

**Comment by fleish** [comment]
I've upgraded a test system to OC7, replaced the file, done occ upgrade and am waiting for occ files:scan --all to complete before checking how things turned out this time.

**Comment by fleish** [comment]
still coming up with the unable to mount shared storage '##' user 'GUID' when trying to access shared folders

**Comment by PVince81** [comment]
Even after unsharing + sharing that one again ? :frowning: 

Looks like this will need some extensive debugging.

**Comment by PVince81** [comment]
Is the owncloud.log still the same as in your first post https://github.com/owncloud/core/issues/12801#issue-51741073 ?

**Comment by fleish** [comment]
I hadn't tried that, but just did and found the following:
- The shares do not show up in the traditional "Shared" folder, only in the new sidebar link "Shared with you"
- When accessing from the new sidebar link "Shared with others", unsharing a folder worked, but resharing it resulted in an error "Sharing somefolder failed, because the file does not exist"
- When accessing the same folder from the previous point in "All files" re-sharing does not result in this error
- Looking at the shared folders in the "All files" it does not show it as shared - therefore it cannot be unshared from there. After unsharing in the "Shared with others" page it can only be re-shared from the All Files page
- When unsharing old folders, owncloud.log repeatedly shows messages like these for all things in the folder:
{"app":"Encryption library","message":"Could not delete shareKey; does not exist: \"\/**REDACTED_GUID1**\/files_encryption\/share-keys\/somefolder\/somefile.shareKey\"","level":3,"time":"2015-01-21T16:35:21+00:00"}

and

{"app":"hook","message":"error while running hook (OCA\\Encryption\\Hooks::postUnshare): Encryption key not found for \"\/**REDACTED_GUID1**\/files\/somefolder\/Workshops\/Security & Privacy.key\/Data\/somefile-4.jpg\" during attempted read via stream","level":3,"time":"2015-01-21T16:35:27+00:00"}

**Comment by PVince81** [comment]
Very strange. It almost looks like the file linked to that share doesn't exist or isn't found for some reason.

Is it the same file as here https://github.com/owncloud/core/issues/12801#issuecomment-70527617 ?
(if you post new DB results, please use the command `.headers on` to make them more readable)

**Comment by fleish** [comment]
I agree it seems like a link is broken somewhere that owncloud half-thinks the file is shared (evidenced by seeing it in the shared with others section but not in the main shared folder seen in the all files section), but when the user it is shared with attempts to access it the system cannot find the file or provide access to it. Perhaps this is because it doesn't think it can mount the shared storage - which I'm still not clear on since it's just a local filesystem.

It looks like when you unshare it goes through each file in the folder and logs the could not delete shareKey message or the Encryption key not found message - but not both. I only pasted one in as there are many files being shared so the number of log messages gets very large.

**Comment by PVince81** [comment]
Is the owner of the file/folder able to access these files (try downloading them) ?

The encryption key message could simply mean that the file is not found.
When unsharing it usually updates the share keys for every file to remove access for the user who is losing access.

**Comment by fleish** [comment]
The owner is able to access them to re-share them. I've rolled back and am re-upgrading to test if they can download.

I think the files must exist because the user is able to re-share them successfully at which point the shared with user can then access them.

As I'm walking through the upgrade steps again, I'm wondering if there is value to running occ files::scan --all earlier in the process. Like before even beginning the upgrade process while still in version 6. Or perhaps after unpacking the version 7 files but prior to issuing the occ upgrade. Not sure if there is any value to either of these ideas.

**Comment by PVince81** [comment]
You mentioned that some of the user's files were broken directly after the upgrade so you could run `occ files:scan --all` on OC 6 before unpacking the 7 files.

There's a chance that those missing files are somehow linked to the shares pointing in the void

**Comment by fleish** [comment]
I tried running occ files::scan --all on OC6 before using apt to reinstall the 7 package files. After the upgrade, I was again unable to see even my own files so I am now re-running occ files::scan --all post upgrade to see if the shares behave correctly after that

**Comment by fleish** [comment]
Still comes up with the same issue when trying to access the shared files via "Shared by others" - unable to mount shared storage ... and nothing shows up in the "All Files -> Shared" folder

**Comment by PVince81** [comment]
@fleish hmmm... I guess something related to the shares is indeed broken. Maybe OC 6 tolerated it but OC 7 is having trouble.

A few points:

1) There is something strange about your database output in https://github.com/owncloud/core/issues/12801#issuecomment-69759608 and https://github.com/owncloud/core/issues/12801#issuecomment-70527617.

I edited the first one to add the missing headers (you should run `.headers on` before running queries on SQLite to make it add headers), and noticed that there was no column `share_type`.
Did you mistakenly remove it when redacting the GUIDs or is that column really missing from your `oc_share` table ? If it was missing, OC shouldn't even be able to work properly.

2) Let's try this on your OC 6 instance: `select * from oc_share where parent is not null and parent not in (select id from oc_share);`.

This is to find child share entry where the "parent" field points to a non-existing share.
This should normally return nothing. If it does, please post the result here. (don't forget `.headers on` :smile:)

3) The SharedStorage is a virtual internal storage that actually translates to the owner's storage. Every shared file/folder is mounted in the recipient's virtual OC filesystem.

4) I suspect that the reason why some user sees zero files in the list is simply because the `list.php` ajax calls fails due to the missing shares, so the file list cannot be returned. (you can see whether `list.php` fails in the network console of the web browser)
The call to `occ files:scan --all` might somehow manage to repair broken shares but I'd recommend not running it for now until we have identified where the breakage is.

**Comment by fleish** [comment]
1) I re-ran the query with headers on and do see the share_type column
id|share_with|uid_owner|parent|item_type|item_source|item_target|file_source|file_target|expiration|token|share_type|permissions|stime|accepted|mail_send
4|ops|**REDACTED_GUID1**||folder|328|/328|328|/Ops|||1|31|1393560988|0|0
6|**REDACTED_GUID2**|**REDACTED_GUID1**||folder|328|/328|328|/Ops|||0|31|1393563973|0|1
7|IT|**REDACTED_GUID1**||folder|356|/356|356|/IT|||1|31|1393605790|0|0
11|**REDACTED_GUID3**|**REDACTED_GUID4**||folder|1199|/1199|1199|/ryand-shared|||0|31|1395760682|0|0

2) Nothing comes back from this query

3) What would cause this to report OC is unable to mount until it is unshared and re-shared? This seems like the crux of the problem given the log messages when trying to view a share in the broken state.

4) I'll need to roll back an instance to OC6 and perform the upgrade to check this before running occ files:scan --all (which only repairs the user's files so they can see them IME).

**Comment by PVince81** [comment]
3) So far I only know of two possible broken shares:
   a) shares that point to a file that doesn't exist any more. We checked that before here https://github.com/owncloud/core/issues/12801#issuecomment-69759608, maybe we should try deleting these bogus entries ? I'm not sure why they would be orphaned at this point.
   b) shares where the parent points to a non-existing share. You said there were none.

4) No need to rollback for now if you have it on OC 7. I thought you were still on OC 6.

**Comment by fleish** [comment]
3a) I can try dropping those in a bit (about to hop a train), but they don't seem to correlate to the share(s) I'm testing and having issues with.

4) My production instance is on 6, but I'm testing on a cloned instance that I keep rolling forward/back. It's currently on 7 as I've been testing proceeding with the upgrade and unsharing/re-sharing. This is not my preferred method, but at some point I may have to cut bait on troubleshooting and live with the headache of having all users with shares go through this.

Is there nothing else the unable to mount storage error would get thrown for in the code?

**Comment by PVince81** [comment]
I think there is a slight chance that deleting these bogus entries will heal the instance. Because OC might try and mount one of them and the failure makes it give up completely and possibly affect anything else.

If you're able to confirm that these orphaned shares were the cause then we might be able to provide a fix to mitigate the issue / or at least make OC ignore them. So far I believed that OC 7 already had such mechanism.

**Comment by fleish** [comment]
I tried deleting orphaned shares from db in OC6, upgrading to OC7, and then checking the web client but the behavior is the same. No personal files and can't access individual shared folders  - I get bumped back to the main page.

**Comment by PVince81** [comment]
Too bad :frowning: 

When you open the "All files" page, do you see an error message in owncloud.log appearing at the same time ? I guess it will probably be the "Shared storage 3 cannot be mounted" one.

Check the browser network console and check the response of the "list.php" call. I expect it should be 500 Internal server error.

Are you able to enable XDebug on the server and see if you can get more details about that "Shared storage 3" error ? (the stack trace)

**Comment by PVince81** [comment]
I think there's one thing we didn't check.
Which user did you login with ? Let's assume it is "GUID1".

Try getting all shares where GUID1 appears as target:
`select * from oc_share where share_with='GUID1'`,

and also as target through a group:
`select * from oc_share where share_with in ('group1', 'group2')`; (replace with the name of the groups the user is in)

Since you deleted the orphaned shares, I expect there is nothing left, but who knows. Let's try anyway:
`select * from oc_filecache where fileid not in (select item_source from oc_share where share_with in ('GUID1', 'group1', 'group2'))`
That should return nothing.

Then check how the oc_filecache entries for these shares look like.
`select * from oc_filecache where fileid in (select item_source from oc_share where share_with in ('GUID1', 'group1', 'group2'))`
That should return nothing.

And after that... the tedious method: manually delete the share entries for that user one by one until the UI shows the files again (and write them down) to find out which one was broken.

**Comment by PVince81** [comment]
Let me know at least when you have the share entries for that user.
Maybe one of the "parent" entries is pointing to NULL instead of the proper parent value (which could be caused by older bugs). When a file/folder is shared with a group, it should normally have one entry with the group share, and sub-entries for each member of the group. Every sub-entry must have its "parent" value point to the main group entry.

**Comment by fleish** [comment]
>When you open the "All files" page, do you see an error message in owncloud.log appearing at the same time ? I guess it will probably be the "Shared storage 3 cannot be mounted" one.

Negative, I don't see anything logged for the blank "All files page" in OC7 (hopefully you meant here and not OC6). The shared storage error only happens when I click on a shared folder within "Shared with you". This also reminds me of another piece of data I may not have mentioned before. The only thing that does show up in "All files" is a folder named "Shared". Pre-OC7 this was the only way to see shared items. Post upgrade, this folder has nothing in it. Only "Shared with you" shows the shared folders. After un-sharing & re-sharing a folder said folder shows up in both locations. I'll also point out the one thing that is changed in the configuration file during the upgrade is the adding of the following: entry in the CONFIG array: 'share_folder' => '/Shared',

>Check the browser network console and check the response of the "list.php" call. I expect it should be 500 Internal server error.

I checked in Chrome's console & network inspectors and I don't see any file named list.php even referenced. I also checked my webserver logs for 500 errors and only found 1 from a few days ago and in relation to a different file.

>Are you able to enable XDebug on the server and see if you can get more details about that "Shared storage 3" error ? (the stack trace)

Can you walk me through how to set this up? I tried installing php5-xdebug, restarting the web server, and adding: define('DEBUG', true); to the bottom of the config.php - outside of the CONFIG array, but nothing new seems to get logged.

I'll try the other things after I grab a bite to eat.

**Comment by PVince81** [comment]
I meant OC 7 all along.

Okay, if users only see the Shared folder then it's not an error when building the list. I was rather expecting an empty list without message (normally it shows "No files here, upload something!")

The file list that you see in the UI is normally based on the files in oc_filecache + the shares shared with that user. Seeing nothing at all, not even the user's own files is strange. Even stranger if the "Shared" folder is the only entry. Did that user have any personal files ?

I wonder if this has something to do with LDAP.
In the data directory every user only has one home, right ?

I suggest to pick a user who had personal files, then unshare everything from that user. If the personal files appear, then this issue is related to shares. If still no files, then unrelated to share and we have a direction to investigate further.

CC @icewind1991 @schiesbn @blizzz in case this rings any bells

**Comment by PVince81** [comment]
About XDebug: when an exception/error is logged it might contain more info. (Ex: the shared storage one)

**Comment by fleish** [comment]
Yes, the user is my own and I definitely have personal files + things shared with others + things shared with me. In "All files" I see just the shared folder and when I enter that folder I see "Nothing in here. Upload something!". Conversely, if I go in through "Shared With Others" I see all the things shared with me, but if I try to access those from that screen I get the spinning wheel for a brief period and ultimately redirected back to "All files" and the unable to mount shared storage message is logged 85 times no matter which shared folder I try to access.

Yes, in the data directory each user has a sub-directory which is their GUID.

I logged in as myself and unshared everything I was sharing to others to no avail. I will try to get other users to unshare things to my user as well.

**Comment by fleish** [comment]
OK, I have nothing shared with me or from me but same behavior - nothing other than the "Shared" folder in "All Files"

**Comment by jcfischer** [comment]
We seem to have run into the same problem when upgrading our production instance yesterday (We did a test with one of the earlier oC7 releases and the upgrade ran fine) and had to roll-back to 6.0.6
I have just skimmed through the comments. I will try to check some of listed DB queries (we are running on Postgres 9.3).

Manually going in and deleting shares however is NOT an option for us: We have around 5500 users on the system and the 2 hour downtime for the upgrade is already bad enough for them...

**Comment by butonic** [comment]
@MorrisJobke Likely a duplicate of https://github.com/owncloud/core/issues/12035. This report also conains relevant log messages.

**Comment by PVince81** [comment]
So does it mean some mechanism would have triggered a LDAP user id format change ?

**Comment by jcfischer** [comment]
we can also provide log files from the tests with a few users. We have seen the `Unable to get mount for shared storage 'X' user` messages during the tests

**Comment by MorrisJobke** [comment]
@jcfischer You said, that an upgrade to an older OC7 worked. Which one was this? 7.0.3? This maybe helps to spot the actual cause of this.

**Comment by jcfischer** [comment]
We did a test upgrade with our (then) live db/filesystem in October/November I think. Either with a 7.0rcSomething or 7.0.1 - I think it was the first Enterprise release we tried

**Comment by MorrisJobke** [comment]
@jcfischer We possibly found the problem: https://github.com/owncloud/core/issues/12035#issuecomment-71993959 We will keep you up to date.

**Comment by jcfischer** [comment]
@MorrisJobke great to hear. We will test the next update before doing it in the productive environment....

**Comment by fleish** [comment]
I can try this on my test environment first if I can get the updated file and confirmation it should be replaced prior to running occ upgrade in OC7.

Also, if someone can tell me how to check if my environment has the same issue for 12035 I could look at that

**Comment by blizzz** [comment]
@fleish you can get the file from https://raw.githubusercontent.com/owncloud/core/fix-12035/apps/user_ldap/appinfo/update.php and replace the corresponding apps/user_ldap/appinfo/update.php before running the upgrade routine.

If I understand correctly, the mentioned bugs are duplicates.

**Comment by fleish** [comment]
I tried the updated file but still seeing the same behavior. Previously, I made backups of the db at each stage. Looking at oc_storages this is what I see:

1) pre-upgrade - running OC6 (providing just a sample here to show unique entries for admin, local, and then my ldap users which run up to the last one shown - 76)
id|numeric_id
home::admin|1
home::**REDACTED_GUID1**|2
home::**REDACTED_GUID2**|3
local::/var/www/owncloud/data/|4
home::**REDACTED_GUID3**|5
.
.
.
home::**REDACTED_GUID4**|76


2) Post OC7 package update & occ upgrade having run. All the previous entries are there up to 76 but are then followed by 57 new entries that don't follow the same pattern
id|numeric_id
home::admin|1
home::**REDACTED_GUID1**|2
home::**REDACTED_GUID2**|3
local::/var/www/owncloud/data/|4
home::**REDACTED_GUID3**|5
.
.
.
home::**REDACTED_GUID4**|76
**Unique 32 character alpha-numeric string without any hypens**|77
**Unique 32 character alpha-numeric string without any hypens**|78
.
.
.
**Unique 32 character alpha-numeric string without any hypens**|133


3) Post running occ files::scan --all to fix personal files not showing up. Again all the previous entries are there up to 133 but they are then followed by 47 new entires that match the original pattern:
id|numeric_id
home::admin|1
home::**REDACTED_GUID1**|2
home::**REDACTED_GUID2**|3
local::/var/www/owncloud/data/|4
home::**REDACTED_GUID3**|5
.
.
.
home::**REDACTED_GUID4**|76
**Unique 32 character alpha-numeric string without any hypens**|77
**Unique 32 character alpha-numeric string without any hypens**|78
.
.
.
**Unique 32 character alpha-numeric string without any hypens**|133
home::**REDACTED_GUID5**|134
.
.
.
home::**REDACTED_GUID6**|151


One final note/plea. Is there any way to convert a pre-OC7 db from sqlite to mysql? I know this wasn't built into occ until OC7, but am wondering how difficult it would be to backport it to OC6. My real pain point that is driving me to upgrade is near-constant DB locking issues. I'd prefer not to have to manually have all my users unshare/re-share files if possible but if I don't sort this out very soon that's the path I'm going to be going down so that we can get to OC7 and use occ to convert the database to fix the constant locking issues we are experiencing. Thanks for the continued help & consideration.

**Comment by fleish** [comment]
Some additional data. Apologies for not thinking to turn on debug through the UI earlier. Here's what I see when I try to open a share via the "Shared with you" section. I'm not pasting every line, but basically I see it reporting path not readable errors for every item it knows about in the data directory - including .ocdata owncloud.log owncloud.db user-directories & so on. Basically everything except for .htaccess I think. I'm not sure what to make of this since it must be able to read these things given their permissions and it's even managing to write to the owncloud.log to report it supposedly cannot read the owncloud.log file:
{"reqId":"54caecb5bc178","app":"user_ldap","message":"readAttribute: uid=me,dc=mydomain,dc=org found","level":0,"time":"2015-01-30T02:30:13+00:00","method":"GET","url":"\/index.php\/apps\/files\/ajax\/list.php?dir=%2FShared%2FIT&sort=name&sortdirection=asc"}
{"reqId":"54caecb5bc178","app":"user_ldap","message":"Requested attribute primarygroupid not found for uid=me,dc=mydomain,dc=org","level":0,"time":"2015-01-30T02:30:14+00:00","method":"GET","url":"\/index.php\/apps\/files\/ajax\/list.php?dir=%2FShared%2FIT&sort=name&sortdirection=asc"}
{"reqId":"54caecb5bc178","app":"user_ldap","message":"readAttribute: uid=sharer1,dc=mydomain,dc=org found","level":0,"time":"2015-01-30T02:30:14+00:00","method":"GET","url":"\/index.php\/apps\/files\/ajax\/list.php?dir=%2FShared%2FIT&sort=name&sortdirection=asc"}
{"reqId":"54caecb5bc178","app":"user_ldap","message":"readAttribute: uid=sharer2,dc=mydomain,dc=org found","level":0,"time":"2015-01-30T02:30:14+00:00","method":"GET","url":"\/index.php\/apps\/files\/ajax\/list.php?dir=%2FShared%2FIT&sort=name&sortdirection=asc"}
{"reqId":"54caecb5bc178","app":"files_sharing","message":"Unable to get mount for shared storage '3' user '**sharer1-GUID**'","level":3,"time":"2015-01-30T02:30:14+00:00","method":"GET","url":"\/index.php\/apps\/files\/ajax\/list.php?dir=%2FShared%2FIT&sort=name&sortdirection=asc"}
{"reqId":"54caecb5bc178","app":"files_sharing","message":"Unable to get mount for shared storage '3' user '**sharer1-GUID**'","level":3,"time":"2015-01-30T02:30:14+00:00","method":"GET","url":"\/index.php\/apps\/files\/ajax\/list.php?dir=%2FShared%2FIT&sort=name&sortdirection=asc"}
{"reqId":"54caecb5bc178","app":"PHP","message":"Invalid argument supplied for foreach() at \/var\/www\/owncloud\/lib\/private\/files\/cache\/scanner.php#216","level":3,"time":"2015-01-30T02:30:14+00:00","method":"GET","url":"\/index.php\/apps\/files\/ajax\/list.php?dir=%2FShared%2FIT&sort=name&sortdirection=asc"}
{"reqId":"54caecb5bc178","app":"files_sharing","message":"Unable to get mount for shared storage '3' user '**sharer1-GUID**'","level":3,"time":"2015-01-30T02:30:14+00:00","method":"GET","url":"\/index.php\/apps\/files\/ajax\/list.php?dir=%2FShared%2FIT&sort=name&sortdirection=asc"}
{"reqId":"54caecb5bc178","app":"OC\\Files\\Cache\\Scanner","message":"!!! Path '**GUID1-of-user-who-is-not-sharing-anything-with-me' is not readable !!!","level":0,"time":"2015-01-30T02:30:14+00:00","method":"GET","url":"\/index.php\/apps\/files\/ajax\/list.php?dir=%2FShared%2FIT&sort=name&sortdirection=asc"}
{"reqId":"54caecb5bc178","app":"files_sharing","message":"Unable to get mount for shared storage '3' user '**sharer1-GUID**'","level":3,"time":"2015-01-30T02:30:14+00:00","method":"GET","url":"\/index.php\/apps\/files\/ajax\/list.php?dir=%2FShared%2FIT&sort=name&sortdirection=asc"}
{"reqId":"54caecb5bc178","app":"OC\\Files\\Cache\\Scanner","message":"!!! Path '**sharer1-GUID**' is not readable !!!","level":0,"time":"2015-01-30T02:30:14+00:00","method":"GET","url":"\/index.php\/apps\/files\/ajax\/list.php?dir=%2FShared%2FIT&sort=name&sortdirection=asc"}
{"reqId":"54caecb5bc178","app":"files_sharing","message":"Unable to get mount for shared storage '3' user '**sharer1-GUID**'","level":3,"time":"2015-01-30T02:30:14+00:00","method":"GET","url":"\/index.php\/apps\/files\/ajax\/list.php?dir=%2FShared%2FIT&sort=name&sortdirection=asc"}
{"reqId":"54caecb5bc178","app":"OC\\Files\\Cache\\Scanner","message":"!!! Path '**GUID2-of-user-who-is-not-sharing-anything-with-me**' is not readable !!!","level":0,"time":"2015-01-30T02:30:14+00:00","method":"GET","url":"\/index.php\/apps\/files\/ajax\/list.php?dir=%2FShared%2FIT&sort=name&sortdirection=asc"}
.
.
.
{"reqId":"54caef724608d","app":"OC\\Files\\Cache\\Scanner","message":"!!! Path '.ocdata' is not readable !!!","level":0,"time":"2015-01-30T02:41:54+00:00","method":"GET","url":"\/index.php\/apps\/files\/ajax\/list.php?dir=%2FShared%2FIT&sort=name&sortdirection=asc"}
.
.
.
{"reqId":"54caef724608d","app":"OC\\Files\\Cache\\Scanner","message":"!!! Path 'owncloud.log' is not
 readable !!!","level":0,"time":"2015-01-30T02:41:54+00:00","method":"GET","url":"\/index.php\/apps\
/files\/ajax\/list.php?dir=%2FShared%2FIT&sort=name&sortdirection=asc"}
.
.
.

**Comment by jcfischer** [comment]
I see the same "Unable to get mount for shared storage" error messages after the failed upgrade, when I click on a shared folder (those are visible to me in the Web UI)

**Comment by MorrisJobke** [comment]
@blizzz Do you have a nice way to get the LDAP config for an OC 6 instance?

**Comment by PVince81** [comment]
@fleish the new GUID entries that you see are the reason for the missing files. It is supposed to reuse the old ones (at least that patch should fix it). As soon as new bogus entries were created it is already too late and would require manual fixing of the database to consolidate the new storage entries with the old ones. The patch should prevent this bogus creation to happen in the first place.
Did you change the advanced LDAP advanced config (uid/guid format) to something else than default ?

@fleish can you post the database is locked error message ? There's a patch for this too for OC7, see https://github.com/owncloud/core/commit/234f33e01e630f763f34c51114d25986bae02b42.patch

**Comment by blizzz** [comment]
> @blizzz Do you have a nice way to get the LDAP config for an OC 6 instance?

```
select configkey, configvalue from oc_appconfig where appid = 'user_ldap';
```

**Comment by fleish** [comment]
@PVince81 I just tested again with the patch and checked the database at each stage and still saw the same behavior. I don't recall what, if anything I may have changed in my LDAP config. Hopefully the DB output below requested by blizzz will answer your Q.

@PVince81 the error happens at different times for different queries, but here is the common part: SQLSTATE[HY000]: General error: 5 database is locked","level":4,"time":"2015-01-30T14:37:28+00:00"}

Here are a few specific ones:
{"app":"webdav","message":"Doctrine\\DBAL\\DBALException: An exception occurred while executing 'DELETE FROM \"oc_permissions\" WHERE \"fileid\" = ?':\n\nSQLSTATE[HY000]: General error: 5 database is locked","level":4,"time":"2015-01-29T17:17:58+00:00"}

{"app":"webdav","message":"Doctrine\\DBAL\\DBALException: An exception occurred while executing 'UPDATE oc_preferences SET configvalue = ? WHERE userid = ? AND appid = ? AND configkey = ?' with params [\"54ca8f5061fcd\", \"**REDACTED_GUID1**\", \"files_sharing\", \"etag\"]:\n\nSQLS
TATE[HY000]: General error: 5 database is locked","level":4,"time":"2015-01-29T19:51:44+00:00"}

{"app":"webdav","message":"Doctrine\\DBAL\\DBALException: An exception occurred while executing 'UPDATE \"oc_filecache\" SET \"mtime\" = ?, \"size\" = ?, \"etag\" = ?, \"storage_mtime\"=? WHERE \"fileid\" = ?':\n\nSQLSTATE[HY000]: General error: 5 database is locked","level":4,"time":"2015-01-29T23:00:55+00:00"}

@blizzz 
configkey|configvalue
enabled|yes
has_memberof_filter_support|
home_folder_naming_rule|
installed_version|0.4.2
last_jpegPhoto_lookup|0
ldap_agent_password|
ldap_attributes_for_group_search|
ldap_attributes_for_user_search|
ldap_backup_host|
ldap_backup_port|
ldap_base|dc=mydomain,dc=org
ldap_base_groups|dc=mydomain,dc=org
ldap_base_users|dc=mydomain,dc=org
ldap_cache_ttl|5
ldap_configuration_active|1
ldap_display_name|cn
ldap_dn|
ldap_email_attr|
ldap_expert_username_attr|
ldap_expert_uuid_group_attr|
ldap_expert_uuid_user_attr|UUID
ldap_group_display_name|cn
ldap_group_filter|(|(cn=owncloud))
ldap_group_filter_mode|0
ldap_group_member_assoc_attribute|memberUid
ldap_groupfilter_groups|
ldap_groupfilter_objectclass|
ldap_host|ldaps://ldap.mydomain.org/
ldap_login_filter|(&(|(objectclass=inetOrgPerson))(uid=%uid))
ldap_login_filter_mode|1
ldap_loginfilter_attributes|
ldap_loginfilter_email|0
ldap_loginfilter_username|1
ldap_nocase|0
ldap_override_main_server|0
ldap_port|636
ldap_quota_attr|
ldap_quota_def|
ldap_tls|0
ldap_turn_off_cert_check|0
ldap_user_filter_mode|1
ldap_userfilter_groups|
ldap_userfilter_objectclass|inetOrgPerson
ldap_userlist_filter|(|(objectclass=inetOrgPerson))
types|authentication

**Comment by blizzz** [comment]
Well, about the config: you did not have the ldap_uuid_attribute value set, so before the patch it would have been set to auto and now it should stay default.

But the DB locked issues seem to be the issue here, i guess, as migration cannot succeed.

**Comment by fleish** [comment]
@blizzz My primary reason for upgrading is to be able to use the occ utility to convert from sqlite to mysql to fix the DB locking issues. If there was a way to do that and stay on OC6 for now I'd jump at the chance as it would allow users to use the system normally. Currently they have ongoing issues when trying to use OC6/sqlite. I saw a previous thread that provided a way to convert DB's but it was from the days of OC5 and there was advice from the OC team not to use it with OC6 due to potential issues.

**Comment by PVince81** [comment]
@fleish I'm on the road at the moment but will look into these "Database is locked" cases next week, the ones that happen during the migration (the patch only fix the "file upload" cases). I have an idea how to fix them. I hope that will also fix your migration issues.

Did you post all the possible error cases ? I need to see the SQL queries that failed to find the right pieces of code that need to look at, so if you have some more different ones, please post them. I will nevertheless try and find all migration related SQL queries and double check them.

Technically the issue happens when a result set cursor is not explicitly closed after a SELECT, and a subsequent UPDATE or DELETE is done on that same table. We need to add explicit "closeCursor" calls to make SQLite happy.

**Comment by PVince81** [comment]
@MorrisJobke are you still working on a script to repair post-damage cases ? I just remembered that the `oc_share` table also contain user ids in `uid_owner` and `share_with`, which means that if user ids need to be manually fixed/renamed, they'd need to be fixed/renamed in that table as well.

Of course, ideally this should not happen in the first place so let's hope we can make the original patch prevent the migration to go wrong.

**Comment by fleish** [comment]
@PVince81 to clarify, the "Database is locked" issue is more of an ongoing problem in OC6 with Sqlite - it is not specific to the migration. My reason for migrating to OC7 is to be able to use the occ tool to convert the database from Sqlite -> Mysql. I'd welcome a way to do this in OC6 to relieve user issues and give me more time to investigate the other migration issues described here (i.e. missing files & broken shares).

**Comment by MorrisJobke** [comment]
@PVince81 Thanks for the reminder. I will have a look at this next week.

**Comment by PVince81** [comment]
@fleish I understand. I hope we can get rid of the "database is locked" issue completely in the upcoming OC 7.0.5. From what I understood above, the LDAP/share issues you are having right now is because the "database is locked" was also happening at a crucial time where the user ids needed to be migrated to a new format/default (@blizzz correct me if I'm wrong). Failure to do so would result in the symptoms you had with missing files / empty homes.

So to sum up, in total there are three problems:
1) Database is locked during LDAP accounts migration (patch to be written)
2) Intermittent "database is locked" during regular use of ownCloud, fixed by patch https://github.com/owncloud/core/commit/234f33e01e630f763f34c51114d25986bae02b42.patch (please apply to your install)
3) Issue with default format for LDAP user ids, fixed by @blizzz's patch https://github.com/owncloud/core/pull/13759.patch (please apply to your install)

**Comment by fleish** [comment]
@PVince81 thanks for confirming as when I read your previous update I thought you were implying the Sqlite DB locked errors were only happening during the migration. I applied the patch in #2 to OC6 and then uploaded a bunch of relatively small files and hit the Sqlite DB locked error on ~70% of them. All except one were on oc_permissions like this:
{"reqId":"54ce7a02df6aa","app":"webdav","message":"Doctrine\\DBAL\\DBALException: An exception occurred while executing 'DELETE FROM \"oc_permissions\" WHERE \"fileid\" = ?':\n\nSQLSTATE[HY000]: General error: 5 database is locked","level":4,"time":"2015-02-01T19:09:54+00:00","method":"PUT","url":"\/remote.php\/webdav\/somedir\/usr\/local\/nagios\/libexec\/check_breeze"}

and just one was for oc_filecache like this:
{"reqId":"54ce7ac4d7290","app":"webdav","message":"Doctrine\\DBAL\\DBALException: An exception occurred while executing 'DELETE FROM \"oc_filecache\" WHERE \"fileid\" = ?':\n\nSQLSTATE[HY000]: General error: 5 database is locked","level":4,"time":"2015-02-01T19:13:08+00:00","method":"PUT","url":"\/remote.php\/webdav\/somedir\/usr\/local\/nagios\/libexec\/eventhandlers\/.svn\/text-base\/fix_domU.sh.svn-base"}

The patch in #3 showed as already applied for me. I tried not applying it and reverse applying it and the Sqlite DB locked errors persisted.

**Comment by PVince81** [comment]
@fleish hmm the patch for file upload works for OC 7 but the code in OC 6 might have other places that suffer from this error (from reading that the error happens on "DELETE", a part that was likely to be modified in OC 7)

The patch number 3 (the LDAP one) doesn't fix any database is locked issue, it is something else. But that something else doesn't kick in **because** of the database is locked issue. It's a bit complicated because it's multiple issues that need to be fixed at the same time.

I'll prepare a fix tomorrow for the migration case. After that hopefully all three fixes/patches can work in unison and make your upgrade to OC 7 work properly. I'll re-post instructions while it's ready.

**Comment by fleish** [comment]
@PVince81 thanks for digging into this. I went to my production instance and found the following UPDATE systems getting blocked by locking as well
{"app":"webdav","message":"Doctrine\\DBAL\\DBALException: An exception occurred while executing 'UPDATE \"oc_filecache\" SET \"mtime\" = ?, \"size\" = ?, \"etag\" = ?, \"storage_mtime\"=? WHERE \"fileid\" = ?':\n\nSQLSTATE[HY000]: General error: 5 database is locked","level":4,"time":"2015-01-28T19:38:58+00:00"}
{"app":"webdav","message":"Doctrine\\DBAL\\DBALException: An exception occurred while executing 'UPDATE \"oc_filecache\" SET \"size\"=? WHERE \"fileid\" = ?':\n\nSQLSTATE[HY000]: General error: 5 database is locked","level":4,"time":"2015-01-28T19:43:33+00:00"}
{"app":"webdav","message":"Doctrine\\DBAL\\DBALException: An exception occurred while executing 'UPDATE \"oc_filecache\" SET \"path_hash\" = ?, \"path\" = ?, \"parent\" = ?, \"name\" = ?, \"mimepart\" = ?, \"mimetype\" = ?, \"size\" = ?, \"mtime\" = ?, \"storage_mtime\" = ?, \"encrypted\" = ?, \"unencrypted_size\" = ?, \"etag\"=? WHERE \"fileid\" = ?':\n\nSQLSTATE[HY000]: General error: 5 database is locked","level":4,"time":"2015-01-29T12:52:02+00:00"}
{"app":"webdav","message":"Doctrine\\DBAL\\DBALException: An exception occurred while executing 'UPDATE \"oc_properties\" SET \"propertypath\" = ? WHERE \"userid\" = ? AND \"propertypath\" = ?':\n\nSQLSTATE[HY000]: General error: 5 database is locked","level":4,"time":"2015-01-29T17:04:27+00:00"}
{"app":"webdav","message":"Doctrine\\DBAL\\DBALException: An exception occurred while executing 'UPDATE oc_preferences SET configvalue = ? WHERE userid = ? AND appid = ? AND configkey = ?' with params [\"54ca8f5061fcd\", \"**REDACTED_GUID**\", \"files_sharing\", \"etag\"]:\n\nSQLSTATE[HY000]: General error: 5 database is locked","level":4,"time":"2015-01-29T19:51:44+00:00"}
{"app":"webdav","message":"Doctrine\\DBAL\\DBALException: An exception occurred while executing 'UPDATE oc_preferences SET configvalue = ? WHERE userid = ? AND appid = ? AND configkey = ?' with params [1422628648, \"**REDACTED_GUID3*\", \"user_ldap\", \"lastJpegPhotoLookup\"]:\n\nSQLSTATE[HY000]: General error: 5 database is locked","level":4,"time":"2015-01-30T14:37:28+00:00"}
{"app":"hook","message":"error while running hook (\\OC\\Files\\Storage\\Shared::setup): An exception occurred while executing 'UPDATE oc_preferences SET configvalue = ? WHERE userid = ? AND appid = ? AND configkey = ?' with params [1422659751, \"**REDACTED_GUID2**\", \"user_ldap\", \"lastJpegPhotoLookup\"]:\n\nSQLSTATE[HY000]: General error: 5 database is locked","level":3,"time":"2015-01-30T23:15:51+00:00"}

**Comment by PVince81** [comment]
@fleish thanks for the detailed info. This was observed on OC 6, right ?
In any cases the fixes will likely to be put into OC 7 so after a successful migration hopefully they'll all disappear. Not sure about backporting to OC 6 yet.

So for the "database is locked" issues we have these now:
- [x] UPDATE oc_filecache with 'size' or 'path_hash': should be covered by https://github.com/owncloud/core/commit/234f33e01e630f763f34c51114d25986bae02b42.patch already
- [ ] Fix "UPDATE oc_preferences"
- [ ] Fix "UPDATE oc_properties"
- [ ] Fix queries in LDAP migration code

**Comment by PVince81** [comment]
- [ ] Double check UPDATE oc_filecache with path_hash though, might happen when moving/renaming files

**Comment by PVince81** [comment]
@blizzz what database changes are done during that LDAP user id migration (update.php from user_ldap?)  apart from changing appconfig values ?

Can you double check your LDAP patch ?

Since it didn't work for @jcfischer either I suspect that the reason why it didn't work for @fleish might be the same (and not directly related to "database is locked" issues).

If that update.php writes to the `oc_preferences` table, then this new patch for oc_preferences in OC 7 could fix it: https://github.com/owncloud/core/pull/13831

**Comment by jcfischer** [comment]
@PVince81 @blizzz: I wonder if you guys shouldn't try to understand what exactly causes the upgrade to fail (i.e. users not seeing the files in the GUI). We are happy to provide excerpts samples of pre and post migration databases - just tell us what you need.

**Comment by fleish** [comment]
@PVince81 that's correct, the locked errors are happening in OC6. FWIW I tried uploading the same files to an instance upgraded to OC7 and did not have get any locked errors in the log. I understand the desire to get everyone on OC7, but I'd also ask to consider the level of effort involved in backporting to OC6 since I think it would benefit a number of users.

**Comment by PVince81** [comment]
@jcfischer The reason why the users don't see files is because they got a new "home" which is empty.

The old install in OC 6 had user ids with a given LDAP user id format, let's call it X. So X(user1) is the id of user1 with format X.

So in the oc_storages table you would see entres with storage id like:
- home::X(user1)
or
- local::/path/to/datadir/X(user1)

And on the filesystem every user has a home folder with the name "X(user1)"

Then in OC 7, there's a bug that resets the uuid format to something else in some cases, let's call that new format Y. So Y(user1) is the user id of user1 with the format Y.

So now when a user "user1" logs in, ownCloud will look for a storage id "home::Y(user1)" in the database (and "local::/path/to/datadir/Y(user1)") But the database only contains storage ids with the format "X(user1)", so it cannot find them and will thus create a new home for these users. So at this point you'll see new entries appear with the storage id:
- home::Y(user1)

and on disk that user will have a new home folder "Y(user1)" using the new format.
Since that new home folder is brand new and empty, the user will not see any files, like a newly created user.

The goal of @blizzz patch is to prevent the format change when upgrading and keep it to what it was before, in your case X(user1). But I don't understand why the patch didn't work for you and is outside my expertise regarding LDAP. I hope @blizzz can help with that part.

Note that @fleish confirmed the format change in his env in this comment https://github.com/owncloud/core/issues/12801#issuecomment-72142954 saying that new entries appeared after upgrade.

Additional note: the "Unique 32 character alpha-numeric string without any hypens" is actually a md5 of the storage id. This happens if the storage id length would exceed 64 characters.

**Comment by fleish** [comment]
@jcfischer can you confirm which earlier 7 release you tested upgrading to without these issues? I'm curious to try this myself and see what comes of it

@pvince81 thanks for confirming that value is an md5 of the storage id

**Comment by fleish** [comment]
@PVince81 @blizzz @jcfischer I managed to poke around and find the following URL: http://download.owncloud.org/community/testing/owncloud-7.0.0RC1.tar.bz2 and gave it a try and low and behold post-upgrade I could see my own files! Shared files were still behaving strangely though. I was able to browse down into my shares past the top level only intermittently. This was an improvement as previously this would fail 100% of the time. I then upgraded to the latest 7.0.4 provided in the repository and things seemed to get much more stable. Shared files now seem to be working reliably 100% of the time. I checked the database post-upgrade and there are no new entries in oc_storages as previously seen.

Hopefully this data point is helpful in figuring out what changed between 7.0.0RC1 & a future release that causes this problem. If nothing else, it may give those of us having a hard time a way forward. I'm going to have some more folks test the resulting instance to make sure it's really working correctly but so far things look very good to me.

Finally, I think it would be beneficial to maintain an easily accessible archive of previous releases & release candidates. I literally had to guess this URL based on something I found in a forum thread.

**Comment by jcfischer** [comment]
Thanks @PVince81 - that makes sense. Weird that the patch didn't help. I did the first upgrade with 7.0.0RC1 as @fleish and we also upgraded to 7.0.3 and 7.0.4 on the test instance. And +1 for the archive of earlier versions, I have also hunted for those. @blizz - if you need anything from me to debug, let me know - I'm happy to re-run the migration on our productive data

**Comment by PVince81** [comment]
Raised ticket for the website to include older versions https://github.com/owncloud/owncloud.org/issues/413

**Comment by blizzz** [comment]
@PVince81 no, LDAP user names do not change upon update. The only thing that is set is a default configuration value for UUID attribute control (which since OC 7 can be set seperately for users and groups).

This does not influence existing users, but has effect on new users and also detecting a changed DN.

**Comment by PVince81** [comment]
Had a discussion with @blizzz and it seems that logically the user id should never change in such situations.

@fleish @jcfischer can you post a snippet from the `oc_ldap_user_mapping` table and see if you observed user id format changes before/after the upgrade ? (you could pick a broken user and find his/her uuid before and after the upgrade). If you don't want to post the result here, please email it to me (pvince81 at owncloud dot com)

From what @fleish mentionned here https://github.com/owncloud/core/issues/12801#issuecomment-72142954 it looks like before the upgrade the uuid was short enough to fit in the format "home::" + uuid. But after the upgrade a few md5 entries started appearing, which means that the uuid might have changed to something that doesn't fit in 32 chars when concatenating "home::" + uuid, which implies a possible unexplained uuid change.

What LDAP server do you guys use ? I guess you didn't switch the LDAP server version/software between the upgrades ?

**Comment by jcfischer** [comment]
@PVince81 I have sent you 100 entries of the db pre / post migration (with the patch applied) - they are identical.

We are using OpenLDAP and no - we didn't change it during the upgrade

**Comment by jcfischer** [comment]
I'm an idiot - I sent you the same data twice... Let me truly unmount the volume with the data, and attach the old database

**Comment by PVince81** [comment]
@jcfischer thanks. Indeed. So the problem must be elsewhere.
Did you too see any new entries popping up in the `oc_storages` table after login ? (related to the users who logged in after the upgrade). Just to confirm that you are having the exact same issue like @fleish.

**Comment by PVince81** [comment]
Ah sorry, didn't see your last message. The "indeed" was for "indeed the data looks the same" :smile:

**Comment by jcfischer** [comment]
@PVince81 the data is still identical though - even with the "real" old data

**Comment by jcfischer** [comment]
@PVince81 the plot thickens: In the post migration oc_storages table I have several thousand new entries in the form: `local::/mnt/data/user.name@example.com` while we had `home::user.name@example.com` in the old database - I have sent you extracts of the data/

**Comment by PVince81** [comment]
Did you guys move the "datadir" during your upgrade ? If yes, please try it again without touching it.
In case you have "local::" entries with hard-coded data dir, it should match exactly whatever is in config.php.

**Comment by PVince81** [comment]
This is not ok. Entries in the format "local::" should not appear for new users since OC >= 5.0.12. New entries would always have the format "home::".
@jcfischer can you confirm that you are not using external storage with type "local" ?

**Comment by jcfischer** [comment]
@PVince81. No - we didn't touch the data directory. We installed the 7.0.4 code and then put back the old configuration file that we are using so that the update process will know that it needs to update the db. 

local storage would be configured where? `mount.json` in the data directory? That is empty

```
root@upgrade-web:/mnt/data# ls -l *json
-rwxrwx--- 1 www-data www-data 2 Jan 27  2014 mount.json
root@upgrade-web:/mnt/data# cat mount.json
[]root@upgrade-web:/mnt/data#
```

**Comment by PVince81** [comment]
Ok thanks. Now we're really in mysterious land.

The code in question is here: https://github.com/owncloud/core/blob/stable7/lib/private/files/filesystem.php#L332

From what I see there are only two conditions that would lead to "Local" storage creation:
1) if `$userObject` is null ?!
2) or if a "local::" storage already exists, then it would use that

So here I wonder if something is causing `$userObject` to become null even though `$user` is set.
@icewind1991 @blizzz please join me on that question.

We have LDAP... what would cause that ? Does it mean that at some point the user manager failed ?

**Comment by PVince81** [comment]
I remember that the only "local::" storage that was left was the anonymous "local::$datadir" storage. All user storages should be in "home::" format. The part in the `else` block should probably pass `null` instead of the `$user` ?

It still doesn't explain why `$userObject` is null in the first place.
We probably need to add additional checks to prevent side-effects when such situation happens.

Needs further research.

**Comment by fleish** [comment]
Does it make sense this issue is not present in 7.0.0RC1 but not is present in 7.0.3 & 7.0.4?

**Comment by PVince81** [comment]
Yes, because in 7.0.0RC1 there were problems with apps update. App update was made more robust later to avoid endless loops and concurrent app updates. So this fits in well with @blizzz's findings.

**Comment by blizzz** [comment]
When we load apps in regular operation, we load first authentication apps, and then filesystem and logging ones, followed by everything else. Maybe we should do the same on upgrade: first upgrade auth apps and enable them, subsequently the same with filesys and logging ones.

**Comment by blizzz** [comment]
@DeepDiver1975 your opinion? ^

**Comment by icewind1991** [comment]
> Maybe we should do the same on upgrade: first upgrade auth apps and enable them, subsequently the same with filesys and logging ones.

Seems like the best way to solve it

**Comment by jcfischer** [comment]
awesome findings. Looking forward to a fix soon. We have planned our next try of the upgrade for March 7th, and I was hoping to test it in the week of February 16th...

**Comment by fleish** [comment]
I'm curious what folks think of the intermediate upgrade plan/path of 6.0.6 -> 7.0.0RC1 -> 7.0.4 (or whatever is current stable at the time) as an alternative?

**Comment by jcfischer** [comment]
@fleish - not that much, actually. The 6.0.6 -> 7.0.4 upgrade process took around 90 minutes of database migration for us. An intermediate update probably won't take that long the second time, but I'd rather limit the total downtime (we need time to test that the first upgrade worked). As it is now, the upgrade process as it is already takes too long for my taste.

**Comment by RobinR1** [comment]
I have encountered a similar problem after upgrade from 7.0.2 to 7.0.4:
2 users of the currently about 520 active cloud-users in our ownCloud environment suddenly received a new entry in oc_storage defined as "local::/srv/www/vhosts/cloud/data/<user-id>". They do already have a "home::<user-id>" entry, but that entry now seems to be ignored. 

More details here: http://forum.owncloud.org/viewtopic.php?f=29&t=26230

**Comment by PVince81** [comment]
There is an experimental pull request here that could solve the issue (at least it seemed to do in my case), but needs more intensive testing before it can be released. If you feel adventurous, here it is: https://github.com/owncloud/core/pull/13865

**Comment by PVince81** [comment]
Note that this PR also requires the other patch https://github.com/owncloud/core/pull/13759.patch.
The patches need to be applied on top of OC 7 **before** running the upgrade routine.

**Comment by blizzz** [comment]
No idea how to test this properly. Too many hard-wired dependencies (OC/Config and OC_App). I can split up doAppUpgrade, but then I am testing two for each loops. Does not feel like a big win. Given, it would detect the issue @PVince81 stumbled upon.

Making the Updater properly testable is a task for 8.1.

**Comment by blizzz** [comment]
Oh, forget it. As talked it makes sense to take out app sorting. OC_App is required for isType check. Unfortunately, it cannot be mocked because it is a static method. We can do things worse, by introducing a second isType method or adding a wrapper or… making it more ugly and error prone.

**Comment by PVince81** [comment]
You probably meant to comment on the PR instead ? https://github.com/owncloud/core/pull/13865

**Comment by blizzz** [comment]
Yupp, sorry 'bout this.

**Comment by PVince81** [comment]
I also raised https://github.com/owncloud/core/pull/14044 to make sure the bogus "local::" storage cannot be created as a side-effect of other bugs.

**Comment by DeepDiver1975** [comment]
@PVince81 is there anything we can do for 7.0.5? Seems rather complex - moving to 7.0.6 ???? THX

**Comment by jcfischer** [comment]
any chance of getting this fixed this week in an official release? We are planning our next upgrade windows on March 7th but need to test it this week. Otherwise we will have to push back the upgrade another 2 weeks

**Comment by PVince81** [comment]
Here is a summary of the current status:

1) If you are having issues with SQLite + database is locked (@fleish the original reporter):
- [x] closeCursor fix, already in stable7/7.0.5 https://github.com/owncloud/core/commit/234f33e01e630f763f34c51114d25986bae02b42.patch
- [ ] additional close cursor for oc_preferences, still needs review for stable7: https://github.com/owncloud/core/pull/14040 (might not be needed but would be safer to have)

2) If you have users that lost files and additional bogus storages suddenly appeared in oc_storages (both @fleish had these too, and you too @jcfischer ?)
- [ ] stable7 patch, still to be reviewed and merged: https://github.com/owncloud/core/pull/14040 (and its master counterpart https://github.com/owncloud/core/pull/13865).
- [x] stable7 patch for LDAP default user id (just in case) is already on stable7: https://github.com/owncloud/core/pull/13759

**Comment by jcfischer** [comment]
@PVince81 - yep, case 2) is our problem from a 6.0.6 to 7.0.4 upgrade

**Comment by DeepDiver1975** [comment]
> closeCursor fix, already in stable7/7.0.5

@PVince81 afaik we reverted that because of issues with oracle - right?

**Comment by LukasReschke** [comment]
I merged No. 2, however, I have no idea if this will get into 7.0.5 or 7.0.6

@karlitschek Any input here?

**Comment by PVince81** [comment]
@DeepDiver1975 dammit, right. Reverted because of Oracle.

**Comment by karlitschek** [comment]
@LukasReschke 7.0.5 is still not build. So this would go into 7.0.5

**Comment by LukasReschke** [comment]
@jcfischer 7.0.5 should work then – or the current stable7 branch.

@karlitschek THX

**Comment by jcfischer** [comment]
@LukasReschke - is 7.0.5 Enterprise going to be built this week?

**Comment by karlitschek** [comment]
@jcfischer Yes. At the same time than 7.0.5 community

**Comment by PVince81** [comment]
@fleish did you have any luck with all the stable7 patches ? See https://github.com/owncloud/core/issues/12801#issuecomment-74668952

**Comment by bboule** [comment]
Hey Guys will this be addressed in 7.0.5?

**Comment by LukasReschke** [comment]
@bboule That question is answered above... But just to rephrase it again: Yes.

**Comment by bboule** [comment]
I thought so... but I need to set expectations with third partied so I wanted to confirm.  Thanks for the help!!

**Comment by fleish** [comment]
@PVince81 sorry, been busy and forgot to respond. I upgraded via the 7.0.0RC1 -> 7.0.4 path and haven't had time to go back and try it with the proper patches.
